### PR TITLE
[WIP] [Stock Market] Add New Cog

### PR DESCRIPTION
**Pull request by zach1502**
### Description of the changes
- Added in a new cog.
- This cog integrates the stock market into Ren using Yahoo Finance's api.
- It's rough around the edges. 100% needs refractoring, and optimizing. but it works.

Also balanced all the casino games so that the expected value is roughly all the same.

### CODE REVIEW

**Review by Injabie3 — CHANGES_REQUESTED** [review]
Please move casino changes into a new PR. It is a separate cog change and should be reviewed separately.

File: cogs/stockmarket/constants.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
+"""Separate file for constants. Cause this takes up too much space."""
+
+class Constants:
```
Comment (Injabie3) [inline cogs/stockmarket/constants.py]: No need for a Constants class, place all your constants outside of the class.

File: cogs/stockmarket/market.py [hunk 2]
```diff
@@ -0,0 +1,568 @@
+""" Here is the Module Docstring"""
```
Comment (Injabie3) [inline cogs/stockmarket/market.py]: Please describe this file in brief or remove this line.

File: cogs/casino/data.py [hunk 3]
```diff
@@ -49,7 +49,7 @@
 
 guild_defaults = {
     "use_old_style": False,
-    "Memberships": {},
```
Comment (Injabie3) [inline cogs/casino/data.py]: Please revert all changes in this file.

File: cogs/casino/games.py [hunk 4]
```diff
@@ -49,6 +49,7 @@ class Core:
         If you roll any other number you must match it on your
         second roll to win.
     """
+    expected_value = 1.14
```
Comment (Injabie3) [inline cogs/casino/games.py]: Please revert all changes in this file.

File: cogs/stockmarket/constants.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    BROKER_NAME = 'Ren\'s Brokerage'
+    DATA_FILE = 'cogs\\stockmarket\\portfolios.json'
+    EMBED_LOCATION = 'cogs\\stockmarket\\chart.jpg'
+    TRANSACTION_CANCELLED_STR = 'Transaction cancelled!'
+    INVALID_TICKER_STR = 'Invalid ticker symbol!'
+    INVALID_AMOUNT_STR = 'Invalid amount!'
+    POOR_STR = 'You do not have enough money to buy this stock!'
+    ACCOUNT_NOT_FOUND_STR = '''No account found! Please create an account using `renmarket createacc`.''' #.format(get prefix somehow)
+    QUIPS = ['Ready to Donate your money to Wall Street I see?',
+    'AAAAAAAAAAND IT\'S GONE!',
```
Comment (Injabie3) [inline cogs/stockmarket/constants.py]: Here and everywhere, would it be worthwhile to use double quotes instead of having to escape single quotes?

File: cogs/stockmarket/market.py [hunk 6]
```diff
@@ -0,0 +1,568 @@
...
+
+                        # order doesn't matter since companies always try to avoid paying out dividends/splits at the same time
+                        # pay out the dividend
+                        shares_owned = holding['amount']
+                        member_object = None  # expected to change
+
+                        # pay dividend
+                        if dividend_due_per_share > 0:
+                            guild_object = self.bot.get_guild(singular_guild_id)
+                            member_object = await guild_object.query_members(user_ids=singular_member_id)
```
Comment (Injabie3) [inline cogs/stockmarket/market.py]: 1. This is a list of member objects, can you please rename this variable to describe it as such?
2. It doesn't look like we're checking for member existence (i.e. member configured a few things beforehand but then leaves the guild). Please add a check to ensure the member object actually exists, and continue to the next member if it doesn't.

File: cogs/stockmarket/market.py [hunk 7]
```diff
@@ -0,0 +1,568 @@
...
+                            'type': 'event',
+                            ticker: [most_recent_unix],
+                            'date': datetime.utcfromtimestamp((int)(time.time())).strftime('%Y-%m-%d %H:%M:%S'),
+                            'dividend_amount': ceil(dividend_due_per_share * shares_owned * 100),
+                            'split_ratio': split_due_per_share
+                        })
+
+    @commands.group()
+    async def market(self, ctx):
+        """Manage your Money! Manage your Life! Invest into the FUTURE!\n
```
Comment (Injabie3) [inline cogs/stockmarket/market.py]: No need for `\n`.

File: cogs/stockmarket/market.py [hunk 8]
```diff
@@ -0,0 +1,568 @@
...
+        into any publicly traded company around the world! #GetRichSlow
+        """
+
+    @market.command(name="sell")
+    async def sell_stock(self, ctx, ticker: str, amount_to_sell: int=0,  user_id: discord.Member=None):
+        """Sell a user-specified amount of a stock"""
+        ticker = ticker.upper()
+
+        if user_id is None:
+            user_id = ctx.author.id
```
Comment (Injabie3) [inline cogs/stockmarket/market.py]: You declared discord.Member for user_id, please use `ctx.author` or change the type hint.

File: cogs/stockmarket/market.py [hunk 9]
```diff
@@ -0,0 +1,568 @@
...
+
+        async with self.config.member(ctx.author).holdings() as holdings:
+            for holding in holdings:
+                if holding['ticker'] == ticker:
+                    amount_owned = holding['amount']
+
+                    while True:
```
Comment (Injabie3) [inline cogs/stockmarket/market.py]: Why do you need an infinite for loop?

File: cogs/stockmarket/market.py [hunk 10]
```diff
@@ -0,0 +1,568 @@
...
+
+                            # append to transactions
+                            async with self.config.member(ctx.author).transactions() as transactions:
+                                transactions.append({
+                                    'type': 'sell',
+                                    'ticker': ticker,
+                                    'date': datetime.utcfromtimestamp((int)(time.time())).strftime('%Y-%m-%d %H:%M:%S'),
+                                    'shares_sold': amount_to_sell,
+                                    'sold_price': price,
+                                    'amount_recieved': amount_to_sell * price
```
Comment (Injabie3) [inline cogs/stockmarket/market.py]: here and in other places with this dict: nit: spelling

File: cogs/stockmarket/market.py [hunk 11]
```diff
@@ -0,0 +1,568 @@
...
+        Hong Kong Stock Exchange (HKSE) = .hk
+        Shenzhen Stock Exchange (SZSE) = .sz
+
+        Tickers for companies may be searched for on https://ca.finance.yahoo.com/
+        Tickers are case-insensitive
+        """
+        ticker = ticker.upper()
+
+        if user_id is None:
+            user_id = ctx.author.id
```
Comment (Injabie3) [inline cogs/stockmarket/market.py]: Same comment as before with reassigning with different type.

File: cogs/stockmarket/market.py [hunk 12]
```diff
@@ -0,0 +1,568 @@
...
+
+        async with self.config.member(ctx.author).holdings() as holdings:
+            if not holdings:
+                await ctx.send(f'{ctx.author.mention} you do not own any shares!')
+                return
+
+            # port_value = 0
+            msg = 'You own the following stocks:\n ```css\n'
+
+            print(holdings)
```
Comment (Injabie3) [inline cogs/stockmarket/market.py]: This prints to console, please remove this or convert to use debug logging.

File: cogs/stockmarket/market.py [hunk 13]
```diff
@@ -0,0 +1,568 @@
...
+
+            for holding in holdings:
+                ticker = holding['ticker']
+                amount = holding['amount']
+                msg += f'[{amount}] shares of [{ticker}]\n'
```
Comment (Injabie3) [inline cogs/stockmarket/market.py]: Consider constructing this first and then using `redbot.core.utils.chat_formatting.box` for backticks

File: cogs/stockmarket/market.py [hunk 14]
```diff
@@ -0,0 +1,568 @@
...
+        await self.get_embed_chart(ticker, period_str, interval_str)
+        await ctx.send(file=discord.File(const.EMBED_LOCATION, filename=ticker + ".jpg"))
+
+    # Simple Information
+    @market.command(name="info")
+    async def get_info(self, ctx, ticker: str):
+        """Gets information about a particular stock"""
+        ticker = ticker.upper()
+        info = yf.Ticker(ticker).info
+        print(info)
```
Comment (Injabie3) [inline cogs/stockmarket/market.py]: Please use `logging.debug` since this prints directly to console.

File: cogs/stockmarket/market.py [hunk 15]
```diff
@@ -0,0 +1,568 @@
...
+
+        embed = await self.build_embed(ctx, ceil(info['regularMarketPrice'] * 100), ticker, info)
+        short_name = info['shortName']
+        embed.set_footer(text=f'{ticker} - {short_name}')
+        try:
+            summary = info['longBusinessSummary']
+            summary = (summary[:1021] + '...') if len(summary) > 1021 else summary
```
Comment (Injabie3) [inline cogs/stockmarket/market.py]: Please make `1021` a constant.

File: cogs/stockmarket/market.py [hunk 16]
```diff
@@ -0,0 +1,568 @@
...
+            msg += '```'
+
+            await ctx.send(msg)
+
+    async def get_embed_chart(self, ticker, period_str='1d', interval_str='5m'):
+        """Gets the chart of the requested stock"""
+        plt.clf()
+        ticker = ticker.upper()
+        # Get the data.. takes like 0.2s
+        data = yf.download(ticker, period=period_str, interval=interval_str)
```
Comment (Injabie3) [inline cogs/stockmarket/market.py]: Here and everywhere we use `yf`, are these blocking call? If so, this could potentially stall the bot.

File: cogs/stockmarket/market.py [hunk 17]
```diff
@@ -0,0 +1,568 @@
...
+        plt.title(ticker)
+        plt.ylabel('Price')
+        plt.grid()
+
+        # convert to webp????
+        plt.savefig(const.EMBED_LOCATION, bbox_inches='tight')  # note: saved as .jpg
+        return True
+
+    # so messy...
+    async def build_embed(self, ctx, price, ticker, ticker_info):
```
Comment (Injabie3) [inline cogs/stockmarket/market.py]: Please add type hints for function args.

**Review by Injabie3 — CHANGES_REQUESTED** [review]
Did one quick scan. Please consider breaking your command coroutine definitions into smaller pieces, since they are hard to follow.

File: cogs/stockmarket/market.py (hunk 16 above)
Comment (zach1502) [reply cogs/stockmarket/market.py]: would taking the `yf` library, and converting it all to async/await functions help unblock?

File: cogs/stockmarket/constants.py [hunk 18]
```diff
@@ -0,0 +1,52 @@
+"""Separate file for constants. Cause this takes up too much space."""
+
+MAX_MSG_LENGTH = 1021
+PRICE_MULTIPLIER = 100
+BROKER_NAME = "Ren's Brokerage"
+DATA_FILE = "cogs\\stockmarket\\portfolios.json"
+EMBED_LOCATION = "cogs\\stockmarket\\chart.jpg"
```
Comment (Injabie3) [inline cogs/stockmarket/constants.py]: Can you please make these paths platform-agnostic? This would work on Windows but not Linux.

File: cogs/stockmarket/constants.py [hunk 19]
```diff
@@ -0,0 +1,52 @@
...
+TRANSACTION_CANCELLED_STR = "Transaction cancelled!"
+INVALID_TICKER_STR = "Invalid ticker symbol!"
+INVALID_AMOUNT_STR = "Invalid amount!"
+POOR_STR = "You do not have enough money to buy this stock!"
+ACCOUNT_NOT_FOUND_STR = """No account found! Please create an account using `renmarket createacc`."""  # .format(get prefix somehow)
```
Comment (Injabie3) [inline cogs/stockmarket/constants.py]: nit: No need for `"""`.

File: cogs/stockmarket/market.py [hunk 20]
```diff
@@ -0,0 +1,554 @@
+from ast import alias
+from math import floor, ceil
+import asyncio
+import time
+import datetime
+import matplotlib.pyplot as plt
+import yfinance as yf  # pip install yfinance
```
Comment (Injabie3) [inline cogs/stockmarket/market.py:6]: Please include a `info.json` and put this as a requirements. You can see other cogs (i.e. YOURLS) for examples.

File: cogs/stockmarket/market.py [hunk 21]
```diff
@@ -0,0 +1,554 @@
...
+        into any publicly traded company around the world! #GetRichSlow
+        """
+        pass
+
+    @market.command(name="debug")
+    async def market_debug(self, ctx: commands.Context):
+        """Debugging command"""
+        # print out all the data
+        all_member_data = await self.config.all_members(guild=None)
+        print(all_member_data)
```
Comment (Injabie3) [inline cogs/stockmarket/market.py]: This should be protected by some mod/admin equivalent check.

File: cogs/stockmarket/market.py [hunk 22]
```diff
@@ -0,0 +1,554 @@
...
+                        value=f"You currently own {amount_owned} shares of {ticker} at {price} {currency_name} each.",
+                        inline=False,
+                    )
+
+                    # sends image as attachment
+                    # thinking of just embeding the image via a link just so it looks prettier
+                    # upload image to imgur first or something idk
+                    await self.get_embed_chart(ticker)
+                    embed_file = discord.File(EMBED_LOCATION, filename=ticker + ".jpg")
+                    embed.set_image(url="attachment://image.jpg")
```
Comment (Injabie3) [inline cogs/stockmarket/market.py]: Does this work if the filename doesn't match the one you're setting with `set_image`?

File: cogs/stockmarket/utils.py [hunk 23]
```diff
@@ -0,0 +1,180 @@
...
+    async def yf_download(self, ticker: str, period_str: str, interval_str: str):
+        return yf.download(ticker, period=period_str, interval=interval_str)
+
+    async def yf_ticker_info(self, ticker: str):
+        return yf.Ticker(ticker).info
+
+    async def yf_ticker_actions(self, ticker: str):
+        return yf.Ticker(ticker).actions
+
+    async def get_embed_chart(self, ticker: str, period_str: str = "1d", interval_str: str = "5m"):
```
Comment (Injabie3) [inline cogs/stockmarket/utils.py]: Where is the input validation for period_str and interval_str? What happens if I enter -2d or 99m?

File: cogs/stockmarket/market.py [hunk 24]
```diff
@@ -0,0 +1,554 @@
...
+
+    async def check_events(self):
+        """Checks if any events are due and pays them out"""
+
+        # Surely there's something better than this
+        all_member_data = await self.config.all_members(guild=None)
+        for singular_guild_id in all_member_data:
+            for singular_member_id in all_member_data[singular_guild_id]:
+                member_dict = all_member_data[singular_guild_id][singular_member_id]
+                for holding in member_dict["holdings"]:
```
Comment (Injabie3) [inline cogs/stockmarket/market.py]: This is a lot of for loops, please split this up into separate functions so that it's easier to read.

File: cogs/stockmarket/utils.py [hunk 25]
```diff
@@ -0,0 +1,180 @@
...
+        embed.add_field(name="52-week high", value=ticker_info["fiftyTwoWeekHigh"], inline=True)
+        embed.add_field(name="Volume", value=ticker_info["volume"], inline=True)
+
+        embed.set_footer(text="Enter 0 to cancel")
+
+        return embed
+
+    async def log_event(
+        self, member_dict: dict, ticker: str, dividend_due: int, split_due_per_share: float
+    ):
```
Comment (Injabie3) [inline cogs/stockmarket/utils.py]: Why is this a coroutine, I don't see other coroutines being used here.

File: cogs/stockmarket/utils.py [hunk 26]
```diff
@@ -0,0 +1,180 @@
...
+        member_dict["transactions"].append(
+            {
+                "type": "event",
+                ticker: ticker,
+                "date": datetime.utcfromtimestamp((int)(time.time())).strftime(
+                    "%Y-%m-%d %H:%M:%S"
```
Comment (Injabie3) [inline cogs/stockmarket/utils.py]: Anyway to simplify this?

File: cogs/stockmarket/utils.py [hunk 27]
```diff
@@ -0,0 +1,180 @@
...
+        plt.grid()
+
+        # convert to webp????
+        plt.savefig(EMBED_LOCATION, bbox_inches="tight")  # note: saved as .jpg
+        return True
+
+    # so messy...
+    async def build_embed(self, ctx: commands.Context, price: int, ticker: str, ticker_info: dict):
+        """Builds the embed"""
+        currency_name = " " + await bank.get_currency_name(ctx.guild)
```
Comment (Injabie3) [inline cogs/stockmarket/utils.py]: Why space at the beginning?

File: cogs/stockmarket/market.py [hunk 28]
```diff
@@ -0,0 +1,554 @@
...
+import yfinance as yf  # pip install yfinance
+import discord
+from datetime import datetime
+
+from redbot.core import bank, commands, Config
+from redbot.core.utils.predicates import MessagePredicate
+from redbot.core.utils.chat_formatting import box
+
+from .constants import *
+from .utils import *
```
Comment (Injabie3) [inline cogs/stockmarket/market.py]: Probably some other places are using `import *`, but can we please explicitly import all attributes that you need? Makes tracing where names are coming from easier.

File: cogs/stockmarket/market.py [hunk 29]
```diff
@@ -0,0 +1,554 @@
...
+        # print out all the data
+        all_member_data = await self.config.all_members(guild=None)
+        print(all_member_data)
+
+    @market.command(name="sell", aliases=["s"])
+    async def sell_stock(
+        self,
+        ctx: commands.Context,
+        ticker: str,
+        amount_to_sell: int = 0,
```
Comment (Injabie3) [inline cogs/stockmarket/market.py]: Any reason not to include default of 0 and let the infra print out the help string if the amount to sell isn't provided?

File: cogs/stockmarket/market.py [hunk 30]
```diff
@@ -0,0 +1,554 @@
...
+        user_id: discord.Member = None,
+    ):
+        """Sell a user-specified amount of a stock"""
+        ticker = ticker.upper()
+
+        if user_id is None:
+            user_id = ctx.author
```
Comment (Injabie3) [inline cogs/stockmarket/market.py]: Re:`user_id`, can we rename this to something that doesn't include ID, since this is not a user ID but rather the Member object itself.

File: cogs/stockmarket/market.py [hunk 31]
```diff
@@ -0,0 +1,554 @@
...
+        Paris Stock Exchange (PSE) = .par
+        London Stock Exchange (LSE) = .l
+
+        Tickers for companies may be searched for on https://ca.finance.yahoo.com/
+        Tickers are case-insensitive
+        """
+        ticker = ticker.upper()
+
+        if user_id is None:
+            user_id = ctx.author
```
Comment (Injabie3) [inline cogs/stockmarket/market.py]: Same as previous sentiment.

File: cogs/stockmarket/utils.py [hunk 32]
```diff
@@ -0,0 +1,180 @@
...
+        plt.ylabel("Price")
+        plt.grid()
+
+        # convert to webp????
+        plt.savefig(EMBED_LOCATION, bbox_inches="tight")  # note: saved as .jpg
+        return True
+
+    # so messy...
+    async def build_embed(self, ctx: commands.Context, price: int, ticker: str, ticker_info: dict):
+        """Builds the embed"""
```
Comment (Injabie3) [inline cogs/stockmarket/utils.py]: nit: does this need to be inside the class? I don't see `self` being used anywhere in this helper.

**Review by Injabie3 — CHANGES_REQUESTED** [review]
Did a second pass, please see below.

**Review by Injabie3 — COMMENTED** [review]
Can you please remove the files at the root of the repo in your PR?
